// weibo/lib-cov/utils.js
/* automatically generated by JSCoverage - do not edit */
if (typeof _$jscoverage === 'undefined') _$jscoverage = {};
if (! _$jscoverage['utils.js']) {
  _$jscoverage['utils.js'] = [];
  _$jscoverage['utils.js'][9] = 0;
  _$jscoverage['utils.js'][10] = 0;
  _$jscoverage['utils.js'][11] = 0;
  _$jscoverage['utils.js'][12] = 0;
  _$jscoverage['utils.js'][17] = 0;
  _$jscoverage['utils.js'][18] = 0;
  _$jscoverage['utils.js'][19] = 0;
  _$jscoverage['utils.js'][28] = 0;
  _$jscoverage['utils.js'][29] = 0;
  _$jscoverage['utils.js'][32] = 0;
  _$jscoverage['utils.js'][33] = 0;
  _$jscoverage['utils.js'][34] = 0;
  _$jscoverage['utils.js'][37] = 0;
  _$jscoverage['utils.js'][40] = 0;
  _$jscoverage['utils.js'][42] = 0;
  _$jscoverage['utils.js'][43] = 0;
  _$jscoverage['utils.js'][44] = 0;
  _$jscoverage['utils.js'][45] = 0;
  _$jscoverage['utils.js'][46] = 0;
  _$jscoverage['utils.js'][47] = 0;
  _$jscoverage['utils.js'][48] = 0;
  _$jscoverage['utils.js'][49] = 0;
  _$jscoverage['utils.js'][51] = 0;
  _$jscoverage['utils.js'][52] = 0;
  _$jscoverage['utils.js'][53] = 0;
  _$jscoverage['utils.js'][56] = 0;
  _$jscoverage['utils.js'][58] = 0;
  _$jscoverage['utils.js'][59] = 0;
  _$jscoverage['utils.js'][60] = 0;
  _$jscoverage['utils.js'][62] = 0;
  _$jscoverage['utils.js'][63] = 0;
  _$jscoverage['utils.js'][64] = 0;
  _$jscoverage['utils.js'][65] = 0;
  _$jscoverage['utils.js'][66] = 0;
  _$jscoverage['utils.js'][68] = 0;
  _$jscoverage['utils.js'][69] = 0;
  _$jscoverage['utils.js'][70] = 0;
  _$jscoverage['utils.js'][72] = 0;
  _$jscoverage['utils.js'][74] = 0;
  _$jscoverage['utils.js'][77] = 0;
  _$jscoverage['utils.js'][78] = 0;
  _$jscoverage['utils.js'][79] = 0;
  _$jscoverage['utils.js'][80] = 0;
  _$jscoverage['utils.js'][82] = 0;
  _$jscoverage['utils.js'][86] = 0;
  _$jscoverage['utils.js'][87] = 0;
  _$jscoverage['utils.js'][88] = 0;
  _$jscoverage['utils.js'][90] = 0;
  _$jscoverage['utils.js'][91] = 0;
  _$jscoverage['utils.js'][93] = 0;
  _$jscoverage['utils.js'][94] = 0;
  _$jscoverage['utils.js'][96] = 0;
  _$jscoverage['utils.js'][98] = 0;
  _$jscoverage['utils.js'][101] = 0;
  _$jscoverage['utils.js'][102] = 0;
  _$jscoverage['utils.js'][103] = 0;
  _$jscoverage['utils.js'][105] = 0;
  _$jscoverage['utils.js'][110] = 0;
  _$jscoverage['utils.js'][111] = 0;
  _$jscoverage['utils.js'][112] = 0;
  _$jscoverage['utils.js'][115] = 0;
  _$jscoverage['utils.js'][116] = 0;
  _$jscoverage['utils.js'][117] = 0;
  _$jscoverage['utils.js'][118] = 0;
  _$jscoverage['utils.js'][119] = 0;
  _$jscoverage['utils.js'][121] = 0;
  _$jscoverage['utils.js'][122] = 0;
  _$jscoverage['utils.js'][125] = 0;
  _$jscoverage['utils.js'][128] = 0;
  _$jscoverage['utils.js'][129] = 0;
  _$jscoverage['utils.js'][130] = 0;
  _$jscoverage['utils.js'][131] = 0;
  _$jscoverage['utils.js'][132] = 0;
  _$jscoverage['utils.js'][134] = 0;
  _$jscoverage['utils.js'][135] = 0;
  _$jscoverage['utils.js'][137] = 0;
  _$jscoverage['utils.js'][138] = 0;
  _$jscoverage['utils.js'][141] = 0;
  _$jscoverage['utils.js'][144] = 0;
  _$jscoverage['utils.js'][145] = 0;
  _$jscoverage['utils.js'][146] = 0;
  _$jscoverage['utils.js'][147] = 0;
  _$jscoverage['utils.js'][148] = 0;
}
_$jscoverage['utils.js'][9]++;
var STRING_FORMAT_REGEX = /\{\{([\w\s\.\'\"\(\),-\[\]]+)?\}\}/g;
_$jscoverage['utils.js'][10]++;
String.prototype.format = (function (values) {
  _$jscoverage['utils.js'][11]++;
  return this.replace(STRING_FORMAT_REGEX, (function (match, key) {
  _$jscoverage['utils.js'][12]++;
  return values[key];
}));
});
_$jscoverage['utils.js'][17]++;
Date.prototype.format = (function (format) {
  _$jscoverage['utils.js'][18]++;
  format = format || "yyyy-MM-dd hh:mm:ss";
  _$jscoverage['utils.js'][19]++;
  var o = {"M+": this.getMonth() + 1, "d+": this.getDate(), "h+": this.getHours(), "m+": this.getMinutes(), "s+": this.getSeconds(), "q+": Math.floor((this.getMonth() + 3) / 3), "S": this.getMilliseconds()};
  _$jscoverage['utils.js'][28]++;
  if (/(y+)/.test(format)) {
    _$jscoverage['utils.js'][29]++;
    format = format.replace(RegExp.$1, (this.getFullYear() + "").substr(4 - RegExp.$1.length));
  }
  _$jscoverage['utils.js'][32]++;
  for (var k in o) {
    _$jscoverage['utils.js'][33]++;
    if (new RegExp("(" + k + ")").test(format)) {
      _$jscoverage['utils.js'][34]++;
      format = format.replace(RegExp.$1, RegExp.$1.length === 1? o[k]: ("00" + o[k]).substr(("" + o[k]).length));
    }
}
  _$jscoverage['utils.js'][37]++;
  return format;
});
_$jscoverage['utils.js'][40]++;
(function () {
  _$jscoverage['utils.js'][42]++;
  var root = this;
  _$jscoverage['utils.js'][43]++;
  var b64_hmac_sha1;
  _$jscoverage['utils.js'][44]++;
  var exports;
  _$jscoverage['utils.js'][45]++;
  var crypto;
  _$jscoverage['utils.js'][46]++;
  if (typeof module === "undefined") {
    _$jscoverage['utils.js'][47]++;
    root.weibo = root.weibo || {};
    _$jscoverage['utils.js'][48]++;
    exports = root.weibo.utils = {};
    _$jscoverage['utils.js'][49]++;
    b64_hmac_sha1 = root.weibo.sha1.b64_hmac_sha1;
  }
  else {
    _$jscoverage['utils.js'][51]++;
    exports = module.exports;
    _$jscoverage['utils.js'][52]++;
    crypto = require("crypto");
    _$jscoverage['utils.js'][53]++;
    b64_hmac_sha1 = require("./sha1").b64_hmac_sha1;
  }
  _$jscoverage['utils.js'][56]++;
  var querystring = {parse: (function (s) {
  _$jscoverage['utils.js'][58]++;
  var qs = {};
  _$jscoverage['utils.js'][59]++;
  if (typeof s !== "string") {
    _$jscoverage['utils.js'][60]++;
    return qs;
  }
  _$jscoverage['utils.js'][62]++;
  var pairs = s.split("&");
  _$jscoverage['utils.js'][63]++;
  for (var i = 0, len = pairs.length; i < len; i++) {
    _$jscoverage['utils.js'][64]++;
    var pair = pairs[i].split("=", 2);
    _$jscoverage['utils.js'][65]++;
    if (pair.length !== 2) {
      _$jscoverage['utils.js'][66]++;
      continue;
    }
    _$jscoverage['utils.js'][68]++;
    var key = pair[0].trim();
    _$jscoverage['utils.js'][69]++;
    if (! key) {
      _$jscoverage['utils.js'][70]++;
      continue;
    }
    _$jscoverage['utils.js'][72]++;
    qs[decodeURIComponent(key)] = decodeURIComponent(pair[1]);
}
  _$jscoverage['utils.js'][74]++;
  return qs;
}), stringify: (function (data) {
  _$jscoverage['utils.js'][77]++;
  var pairs = [];
  _$jscoverage['utils.js'][78]++;
  data = data || {};
  _$jscoverage['utils.js'][79]++;
  for (var k in data) {
    _$jscoverage['utils.js'][80]++;
    pairs.push(encodeURIComponent(k) + "=" + encodeURIComponent("" + data[k]));
}
  _$jscoverage['utils.js'][82]++;
  return pairs.join("&");
})};
  _$jscoverage['utils.js'][86]++;
  function urljoin(url, params) {
    _$jscoverage['utils.js'][87]++;
    if (typeof params === "object") {
      _$jscoverage['utils.js'][88]++;
      params = querystring.stringify(params);
    }
    _$jscoverage['utils.js'][90]++;
    if (! params) {
      _$jscoverage['utils.js'][91]++;
      return url;
    }
    _$jscoverage['utils.js'][93]++;
    if (url.indexOf("?") < 0) {
      _$jscoverage['utils.js'][94]++;
      url += "?";
    }
    else {
      _$jscoverage['utils.js'][96]++;
      url += "&";
    }
    _$jscoverage['utils.js'][98]++;
    return url + params;
}
  _$jscoverage['utils.js'][101]++;
  function base64HmacSha1(baseString, key) {
    _$jscoverage['utils.js'][102]++;
    if (b64_hmac_sha1) {
      _$jscoverage['utils.js'][103]++;
      return b64_hmac_sha1(key, baseString);
    }
    _$jscoverage['utils.js'][105]++;
    return new crypto.Hmac().init("sha1", key).update(baseString).digest("base64");
}
  _$jscoverage['utils.js'][110]++;
  function htmlencode(str) {
    _$jscoverage['utils.js'][111]++;
    if (! str) {
      _$jscoverage['utils.js'][111]++;
      return "";
    }
    _$jscoverage['utils.js'][112]++;
    return str.replace(/&/g, "&amp;").replace(/</g, "&lt;").replace(/>/g, "&gt;");
}
  _$jscoverage['utils.js'][115]++;
  exports.extend = (function (destination) {
  _$jscoverage['utils.js'][116]++;
  for (var i = 1, len = arguments.length; i < len; i++) {
    _$jscoverage['utils.js'][117]++;
    var source = arguments[i];
    _$jscoverage['utils.js'][118]++;
    if (! source) {
      _$jscoverage['utils.js'][119]++;
      continue;
    }
    _$jscoverage['utils.js'][121]++;
    for (var property in source) {
      _$jscoverage['utils.js'][122]++;
      destination[property] = source[property];
}
}
  _$jscoverage['utils.js'][125]++;
  return destination;
});
  _$jscoverage['utils.js'][128]++;
  exports.inherits = (function (destination) {
  _$jscoverage['utils.js'][129]++;
  for (var i = 1, len = arguments.length; i < len; i++) {
    _$jscoverage['utils.js'][130]++;
    var source = arguments[i];
    _$jscoverage['utils.js'][131]++;
    if (! source) {
      _$jscoverage['utils.js'][132]++;
      continue;
    }
    _$jscoverage['utils.js'][134]++;
    for (var property in source) {
      _$jscoverage['utils.js'][135]++;
      destination[property] = source[property];
}
    _$jscoverage['utils.js'][137]++;
    if (destination.super_ === undefined) {
      _$jscoverage['utils.js'][138]++;
      destination.super_ = source;
    }
}
  _$jscoverage['utils.js'][141]++;
  return destination;
});
  _$jscoverage['utils.js'][144]++;
  exports.STRING_FORMAT_REGEX = STRING_FORMAT_REGEX;
  _$jscoverage['utils.js'][145]++;
  exports.querystring = querystring;
  _$jscoverage['utils.js'][146]++;
  exports.base64HmacSha1 = base64HmacSha1;
  _$jscoverage['utils.js'][147]++;
  exports.urljoin = urljoin;
  _$jscoverage['utils.js'][148]++;
  exports.htmlencode = htmlencode;
})();
_$jscoverage['utils.js'].source = ["/**"," * &#230;&#160;&#188;&#229;&#188;&#143;&#229;&#140;&#150;&#229;&#173;&#151;&#231;&#172;&#166;&#228;&#184;&#178;"," * eg:"," *  '{0}&#229;&#164;&#169;&#230;&#156;&#137;{1}&#228;&#184;&#170;&#229;&#176;&#143;&#230;&#151;&#182;'.format([1, 24]) "," *  or"," *  '{{day}}&#229;&#164;&#169;&#230;&#156;&#137;{{hour}}&#228;&#184;&#170;&#229;&#176;&#143;&#230;&#151;&#182;'.format({day:1, hour:24}})"," * @param {Object} values"," */","var STRING_FORMAT_REGEX = /\\{\\{([\\w\\s\\.\\'\\\"\\(\\),-\\[\\]]+)?\\}\\}/g;","String.prototype.format = function (values) {","  return this.replace(STRING_FORMAT_REGEX, function(match, key) {","    return values[key];","  });","};","","// &#230;&#160;&#188;&#229;&#188;&#143;&#229;&#140;&#150;&#230;&#151;&#182;&#233;&#151;&#180;&#232;&#190;&#147;&#229;&#135;&#186;&#227;&#128;&#130;&#231;&#164;&#186;&#228;&#190;&#139;&#239;&#188;&#154;new Date().format(\"yyyy-MM-dd hh:mm:ss\");","Date.prototype.format = function (format) {","  format = format || \"yyyy-MM-dd hh:mm:ss\";","  var o = {","    \"M+\" : this.getMonth() + 1, //month","    \"d+\" : this.getDate(),    //day","    \"h+\" : this.getHours(),   //hour","    \"m+\" : this.getMinutes(), //minute","    \"s+\" : this.getSeconds(), //second","    \"q+\" : Math.floor((this.getMonth() + 3) / 3), //quarter","    \"S\" : this.getMilliseconds() //millisecond","  };","  if (/(y+)/.test(format)) {","    format = format.replace(RegExp.$1, (this.getFullYear() + \"\").substr(4 - RegExp.$1.length));","  }","","  for (var k in o) {","    if (new RegExp(\"(\"+ k +\")\").test(format)) {","      format = format.replace(RegExp.$1, RegExp.$1.length === 1 ? o[k] : (\"00\" + o[k]).substr((\"\" + o[k]).length));","    }","  }","  return format;","};","","(function () {","  ","  var root = this; // window on browser","  var b64_hmac_sha1;","  var exports;","  var crypto;","  if (typeof module === 'undefined') {","    root.weibo = root.weibo || {};","    exports = root.weibo.utils = {};","    b64_hmac_sha1 = root.weibo.sha1.b64_hmac_sha1;","  } else {","    exports = module.exports;","    crypto = require('crypto');","    b64_hmac_sha1 = require('./sha1').b64_hmac_sha1;","  }","","  var querystring = {","    parse: function (s) {","      var qs = {};","      if (typeof s !== 'string') {","        return qs;","      }","      var pairs = s.split('&amp;');","      for (var i = 0, len = pairs.length; i &lt; len; i++) {","        var pair = pairs[i].split('=', 2);","        if (pair.length !== 2) {","          continue;","        }","        var key = pair[0].trim();","        if (!key) {","          continue;","        }","        qs[decodeURIComponent(key)] = decodeURIComponent(pair[1]);","      }","      return qs;","    },","    stringify: function (data) {","      var pairs = [];","      data = data || {};","      for (var k in data) {","        pairs.push(encodeURIComponent(k) + '=' + encodeURIComponent('' + data[k]));","      }","      return pairs.join('&amp;');","    }","  };","","  function urljoin(url, params) {","    if (typeof params === 'object') {","      params = querystring.stringify(params);","    }","    if (!params) {","      return url;","    }","    if (url.indexOf('?') &lt; 0) {","      url += '?';","    } else {","      url += '&amp;';","    }","    return url + params;","  }","","  function base64HmacSha1(baseString, key) {","    if (b64_hmac_sha1) {","      return b64_hmac_sha1(key, baseString);","    }","    return new crypto.Hmac().init(\"sha1\", key).update(baseString).digest(\"base64\");","  }","","  // HTML &#231;&#188;&#150;&#231;&#160;&#129;","  // test: hard code testing &#227;&#128;&#130;&#227;&#128;&#130;&#227;&#128;&#130; '\"!@#$%^&amp;*()-=+ |][ {} ~` &amp;&amp;&amp;&amp;&amp;amp; &amp;lt; &amp; C++ c++c + +c &amp;amp;","  function htmlencode(str) {","    if (!str) { return ''; }","    return str.replace(/&amp;/g, '&amp;amp;').replace(/&lt;/g, '&amp;lt;').replace(/&gt;/g, '&amp;gt;');","  }","","  exports.extend = function (destination) {","    for (var i = 1, len = arguments.length; i &lt; len; i++) {","      var source = arguments[i];","      if (!source) {","        continue;","      }","      for (var property in source) {","        destination[property] = source[property];","      }","    }","    return destination;","  };","","  exports.inherits = function (destination) {","    for (var i = 1, len = arguments.length; i &lt; len; i++) {","      var source = arguments[i];","      if (!source) {","        continue;","      }","      for (var property in source) {","        destination[property] = source[property];","      }","      if (destination.super_ === undefined) {","        destination.super_ = source;","      }","    }","    return destination;","  };","","  exports.STRING_FORMAT_REGEX = STRING_FORMAT_REGEX;","  exports.querystring = querystring;","  exports.base64HmacSha1 = base64HmacSha1;","  exports.urljoin = urljoin;","  exports.htmlencode = htmlencode;","","})();"];
